libs/lb/lb-rs/src/svg/buffer.rs

```diff
@@ -0,0 +1,482 @@
+use std::fmt::Write;
+use std::sync::Arc;
+use std::{collections::HashMap, str::FromStr};
+
+use crate::blocking::Lb;
+use crate::model::file_metadata::DocumentHmac;
+
+use bezier_rs::{Bezier, Subpath};
+use glam::{DAffine2, DMat2, DVec2};
+use indexmap::IndexMap;
+use usvg::{
+    fontdb::Database,
+    tiny_skia_path::{PathSegment, Point},
+    ImageHrefResolver, ImageHrefStringResolverFn, Options, Transform,
+};
+use usvg::{Color, ImageKind, Paint};
+use uuid::Uuid;
+
+use super::element::{DynamicColor, Stroke};
+use super::{
+    diff::DiffState,
+    element::{Element, Image, ManipulatorGroupId, Path},
+};
+
+const ZOOM_G_ID: &str = "lb_master_transform";
+
+#[derive(Default)]
+pub struct Buffer {
+    pub open_file_hmac: Option<DocumentHmac>,
+    pub opened_content: String,
+
+    pub elements: IndexMap<Uuid, Element>,
+    pub master_transform: Transform,
+    id_map: HashMap<Uuid, String>,
+}
+
+impl Buffer {
+    pub fn new(
+        content: &str, maybe_core: Option<&Lb>, open_file_hmac: Option<DocumentHmac>,
+    ) -> Self {
+        let mut elements = IndexMap::default();
+        let mut master_transform = Transform::identity();
+        let mut id_map = HashMap::default();
+
+        let opt = if let Some(core) = maybe_core {
+            let lb_local_resolver = ImageHrefResolver {
+                resolve_data: ImageHrefResolver::default_data_resolver(),
+                resolve_string: lb_local_resolver(core),
+            };
+            Options { image_href_resolver: lb_local_resolver, ..Default::default() }
+        } else {
+            Options::default()
+        };
+
+        let maybe_tree = usvg::Tree::from_str(content, &opt, &Database::default());
+
+        if let Err(err) = maybe_tree {
+            println!("{:#?}", err);
+            println!("couldn't parse the base content");
+        } else {
+            let utree = maybe_tree.unwrap();
+
+            utree
+                .root()
+                .children()
+                .iter()
+                .enumerate()
+                .for_each(|(_, u_el)| {
+                    parse_child(u_el, &mut elements, &mut master_transform, &mut id_map)
+                });
+        }
+
+        Self {
+            open_file_hmac,
+            opened_content: content.to_string(),
+            elements,
+            master_transform,
+            id_map,
+        }
+    }
+
+    pub fn reload(
+        local_elements: &mut IndexMap<Uuid, Element>, local_master_transform: Transform,
+        base_content: &str, remote_content: &str,
+    ) {
+        let base_buffer = Buffer::new(base_content, None, None);
+
+        let remote_buffer = Buffer::new(remote_content, None, None);
+
+        for (id, base_el) in base_buffer.elements.iter() {
+            if let Some(remote_el) = remote_buffer.elements.get(id) {
+                if remote_el != base_el {
+                    // this element was changed remotly
```

Comment:
nit: spelling errors in comments across this file (or is this how gen X spells 'remotely'?)